<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta name="description" content="">
    <meta name="author" content="">

    <title>Reproducibility Guide</title>

    <!-- Bootstrap core CSS -->
    <link href="/assets/css/bootstrap.css" rel="stylesheet">

    <!-- Custom styles for this template -->
    <link href="/assets/css/main.css" rel="stylesheet">
    <link href="/assets/css/font-awesome.min.css" rel="stylesheet">

    <script src="https://code.jquery.com/jquery-1.10.2.min.js"></script>
    <script src="/assets/js/chart.js"></script>


    <!-- HTML5 shim and Respond.js IE8 support of HTML5 elements and media queries -->
    <!--[if lt IE 9]>
      <script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
      <script src="https://oss.maxcdn.com/libs/respond.js/1.3.0/respond.min.js"></script>
    <![endif]-->
  
   <style>
  body {margin-top: 100px;} /*This is the margin from the navigation to the next object.*/
  </style>

  </head>

  <body>

    <!-- Fixed navbar -->
    <div class="navbar navbar-default navbar-fixed-top">
      <div class="container">
        <div class="navbar-header">
          <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".navbar-collapse">
            <span class="icon-bar"></span>
            <span class="icon-bar"></span>
            <span class="icon-bar"></span>
          </button>
          <a class="navbar-brand" href="/"><i class="fa fa-book"></i></a>
        </div>
        <div class="navbar-collapse collapse">
          <ul class="nav navbar-nav navbar-right">
            <li class="active"><a href="https://github.com/ropensci/reproducibility-guide">Contribute <i class="fa fa-github"></i></a></li>
          </ul>
        </div><!--/.nav-collapse -->
      </div>
    </div>

    <div id="green">
        <div class="container">
            <div class="col-lg-8 col-lg-offset-1">
                <div class="row">
            
                   
                <h1><b> How to share data with a statistician </b></h1>
                </br>
                </br>
                
                    
                </div>
            </div>
        </div>
    </div>

<div id="white">
        <div class="container">
            <div class="col-lg-10 col-lg-offset-1">
                <div class="row">
            
                </br>
                </br>

                <p><em>The origin of this content is located at <a href="https://github.com/jtleek/datasharing">https://github.com/jtleek/datasharing</a></em></p>

<p>This is a guide for anyone who needs to share data with a statistician. The target audiences I have in mind are:</p>

<ul>
<li>Scientific collaborators who need statisticians to analyze data for them</li>
<li>Students or postdocs in scientific disciplines looking for consulting advice</li>
<li>Junior statistics students whose job it is to collate/clean data sets</li>
</ul>

<p>The goals of this guide are to provide some instruction on the best way to share data to avoid the most common pitfalls
and sources of delay in the transition from data collection to data analysis. The <a href="http://biostat.jhsph.edu/%7Ejleek/">Leek group</a> works with a large
number of collaborators and the number one source of variation in the speed to results is the status of the data
when they arrive at the Leek group. Based on my conversations with other statisticians this is true nearly universally.</p>

<p>My strong feeling is that statisticians should be able to handle the data in whatever state they arrive. It is important
to see the raw data, understand the steps in the processing pipeline, and be able to incorporate hidden sources of
variability in one&#39;s data analysis. On the other hand, for many data types, the processing steps are well documented
and standardized. So the work of converting the data from raw form to directly analyzable form can be performed 
before calling on a statistician. This can dramatically speed the turnaround time, since the statistician doesn&#39;t
have to work through all the pre-processing steps first. </p>

<h1>What you should deliver to the statistician</h1>

<p>For maximum speed in the analysis this is the information you should pass to a statistician:</p>

<ol>
<li>The raw data.</li>
<li>A <a href="http://vita.had.co.nz/papers/tidy-data.pdf">tidy data set</a> </li>
<li>A code book describing each variable and its values in the tidy data set.<br></li>
<li>An explicit and exact recipe you used to go from 1 -&gt; 2,3 </li>
</ol>

<p>Let&#39;s look at each part of the data package you will transfer. </p>

<h3>The raw data</h3>

<p>It is critical that you include the rawest form of the data that you have access to. Here are some examples of the
raw form of data:</p>

<ul>
<li>The strange <a href="http://en.wikipedia.org/wiki/Binary_file">binary file</a> your measurement machine spits out</li>
<li>The unformatted Excel file with 10 worksheets the company you contracted with sent you</li>
<li>The complicated <a href="http://en.wikipedia.org/wiki/JSON">JSON</a> data you got from scraping the <a href="https://twitter.com/twitterapi">Twitter API</a></li>
<li>The hand-entered numbers you collected looking through a microscope</li>
</ul>

<p>You know the raw data is in the right format if you: </p>

<ol>
<li>Ran no software on the data</li>
<li>Did not manipulate any of the numbers in the data</li>
<li>You did not remove any data from the data set</li>
<li>You did not summarize the data in any way</li>
</ol>

<p>If you did any manipulation of the data at all it is not the raw form of the data. Reporting manipulated data
as raw data is a very common way to slow down the analysis process, since the analyst will often have to do a
forensic study of your data to figure out why the raw data looks weird. </p>

<h3>The tidy data set</h3>

<p>The general principles of tidy data are laid out by <a href="http://had.co.nz/">Hadley Wickham</a> in <a href="http://vita.had.co.nz/papers/tidy-data.pdf">this paper</a>
and <a href="http://vimeo.com/33727555">this video</a>. The paper and the video are both focused on the <a href="http://www.r-project.org/">R</a> package, which you
may or may not know how to use. Regardless the four general principles you should pay attention to are:</p>

<ol>
<li>Each variable you measure should be in one column</li>
<li>Each different observation of that variable should be in a different row</li>
<li>There should be one table for each &quot;kind&quot; of variable</li>
<li>If you have multiple tables, they should include a column in the table that allows them to be linked</li>
</ol>

<p>While these are the hard and fast rules, there are a number of other things that will make your data set much easier
to handle. First is to include a row at the top of each data table/spreadsheet that contains full row names. 
So if you measured age at diagnosis for patients, you would head that column with the name <code>AgeAtDiagnosis</code> instead
of something like <code>ADx</code> or another abbreviation that may be hard for another person to understand. </p>

<p>Here is an example of how this would work from genomics. Suppose that for 20 people you have collected gene expression measurements with 
<a href="http://en.wikipedia.org/wiki/RNA-Seq">RNA-sequencing</a>. You have also collected demographic and clinical information
about the patients including their age, treatment, and diagnosis. You would have one table/spreadsheet that contains the clinical/demographic
information. It would have four columns (patient id, age, treatment, diagnosis) and 21 rows (a row with variable names, then one row
for every patient). You would also have one spreadsheet for the summarized genomic data. Usually this type of data
is summarized at the level of the number of counts per exon. Suppose you have 100,000 exons, then you would have a
table/spreadsheet that had 21 rows (a row for gene names, and one row for each patient) and 100,001 columns (one row for patient
ids and one row for each data type). </p>

<p>If you are sharing your data with the collaborator in Excel, the tidy data should be in one Excel file per table. They
should not have multiple worksheets, no macros should be applied to the data, and no columns/cells should be highlighted. 
Alternatively share the data in a <a href="http://en.wikipedia.org/wiki/Comma-separated_values">CSV</a> or <a href="http://en.wikipedia.org/wiki/Tab-separated_values">TAB-delimited</a> text file.</p>

<h3>The code book</h3>

<p>For almost any data set, the measurements you calculate will need to be described in more detail than you will sneak
into the spreadsheet. The code book contains this information. At minimum it should contain:</p>

<ol>
<li>Information about the variables (including units!) in the data set not contained in the tidy data </li>
<li>Information about the summary choices you made</li>
<li>Information about the experimental study design you used</li>
</ol>

<p>In our genomics example, the analyst would want to know what the unit of measurement for each
clinical/demographic variable is (age in years, treatment by name/dose, level of diagnosis and how heterogeneous). They 
would also want to know how you picked the exons you used for summarizing the genomic data (UCSC/Ensembl, etc.). They
would also want to know any other information about how you did the data collection/study design. For example,
are these the first 20 patients that walked into the clinic? Are they 20 highly selected patients by some characteristic
like age? Are they randomized to treatments? </p>

<p>A common format for this document is a Word file. There should be a section called &quot;Study design&quot; that has a thorough
description of how you collected the data. There is a section called &quot;Code book&quot; that describes each variable and its
units. </p>

<h3>How to code variables</h3>

<p>When you put variables into a spreadsheet there are several main categories you will run into depending on their <a href="http://en.wikipedia.org/wiki/Statistical_data_type">data type</a>:</p>

<ol>
<li>Continuous</li>
<li>Ordinal</li>
<li>Categorical</li>
<li>Missing </li>
<li>Censored</li>
</ol>

<p>Continuous variables are anything measured on a quantitative scale that could be any fractional number. An example
would be something like weight measured in kg. <a href="http://en.wikipedia.org/wiki/Ordinal_data">Ordinal data</a> are data that have a fixed, small (&lt; 100) number of levels but are ordered. 
This could be for example survey responses where the choices are: poor, fair, good. <a href="http://en.wikipedia.org/wiki/Categorical_variable">Categorical data</a> are data where there
are multiple categories, but they aren&#39;t ordered. One example would be sex: male or female. <a href="http://en.wikipedia.org/wiki/Missing_data">Missing data</a> are data
that are missing and you don&#39;t know the mechanism. You should code missing values as <code>NA</code>. <a href="http://en.wikipedia.org/wiki/Censoring_(statistics)">Censored data</a> are data
where you know the missingness mechanism on some level. Common examples are a measurement being below a detection limit
or a patient being lost to follow-up. They should also be coded as <code>NA</code> when you don&#39;t have the data. But you should
also add a new column to your tidy data called, &quot;VariableNameCensored&quot; which should have values of <code>TRUE</code> if censored 
and <code>FALSE</code> if not. In the code book you should explain why those values are missing. It is absolutely critical to report
to the analyst if there is a reason you know about that some of the data are missing. You should also not <a href="http://en.wikipedia.org/wiki/Imputation_(statistics)">impute</a>/make up/
throw away missing observations.</p>

<p>In general, try to avoid coding categorical or ordinal variables as numbers. When you enter the value for sex in the tidy
data, it should be &quot;male&quot; or &quot;female&quot;. The ordinal values in the data set should be &quot;poor&quot;, &quot;fair&quot;, and &quot;good&quot; not 1, 2 ,3.
This will avoid potential mixups about which direction effects go and will help identify coding errors. </p>

<p>Always encode every piece of information about your observations using text. For example, if you are storing data in Excel and use a form of colored text or cell background formatting to indicate information about an observation (&quot;red variable entries were observed in experiment 1.&quot;) then this information will not be exported (and will be lost!) when the data is exported as raw text.  Every piece of data should be encoded as actual text that can be exported.  </p>

<h3>The instruction list/script</h3>

<p>You may have heard this before, but <a href="http://www.sciencemag.org/content/334/6060/1226">reproducibility is kind of a big deal in computational science</a>.
That means, when you submit your paper, the reviewers and the rest of the world should be able to exactly replicate
the analyses from raw data all the way to final results. If you are trying to be efficient, you will likely perform
some summarization/data analysis steps before the data can be considered tidy. </p>

<p>The ideal thing for you to do when performing summarization is to create a computer script (in <code>R</code>, <code>Python</code>, or something else) 
that takes the raw data as input and produces the tidy data you are sharing as output. You can try running your script
a couple of times and see if the code produces the same output. </p>

<p>In many cases, the person who collected the data has incentive to make it tidy for a statistician to speed the process
of collaboration. They may not know how to code in a scripting language. In that case, what you should provide the statistician
is something called <a href="http://en.wikipedia.org/wiki/Pseudocode">pseudocode</a>. It should look something like:</p>

<ol>
<li>Step 1 - take the raw file, run version 3.1.2 of summarize software with parameters a=1, b=2, c=3</li>
<li>Step 2 - run the software separately for each sample</li>
<li>Step 3 - take column three of outputfile.out for each sample and that is the corresponding row in the output data set</li>
</ol>

<p>You should also include information about which system (Mac/Windows/Linux) you used the software on and whether you 
tried it more than once to confirm it gave the same results. Ideally, you will run this by a fellow student/labmate
to confirm that they can obtain the same output file you did. </p>

<h1>What you should expect from the analyst</h1>

<p>When you turn over a properly tidied data set it dramatically decreases the workload on the statistician. So hopefully
they will get back to you much sooner. But most careful statisticians will check your recipe, ask questions about
steps you performed, and try to confirm that they can obtain the same tidy data that you did with, at minimum, spot
checks.</p>

<p>You should then expect from the statistician:</p>

<ol>
<li>An analysis script that performs each of the analyses (not just instructions)</li>
<li>The exact computer code they used to run the analysis</li>
<li>All output files/figures they generated. </li>
</ol>

<p>This is the information you will use in the supplement to establish reproducibility and precision of your results. Each
of the steps in the analysis should be clearly explained and you should ask questions when you don&#39;t understand
what the analyst did. It is the responsibility of both the statistician and the scientist to understand the statistical
analysis. You may not be able to perform the exact analyses without the statistician&#39;s code, but you should be able
to explain why the statistician performed each step to a labmate/your principal investigator. </p>

<h1>Contributors</h1>

<ul>
<li><a href="http://biostat.jhsph.edu/%7Ejleek/">Jeff Leek</a> - Wrote the initial version.</li>
<li><a href="http://bit.ly/LColladoTorres">L. Collado-Torres</a> - Fixed typos, added links.</li>
<li><a href="http://people.umass.edu/nick/">Nick Reich</a> - Added tips on storing data as text.</li>
</ul>

                    
                </br>
                </br>

                </div>
            </div>
        </div>
    </div>


<div id="white">
<div class="container">
    <div class="row">
        <div class="col-lg-10 col-lg-offset-1">
            <div id="disqus_thread"></div>
                <script type="text/javascript">
                 /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
                var disqus_shortname = 'reproducibilityguide'; // required: replace example with your forum shortname

                /* * * DON'T EDIT BELOW THIS LINE * * */
                (function() {
                    var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
                    dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
                    (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
                })();
                </script>
            <noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
            <a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>

       </div><!-- /col-lg-8 -->
    </div><!-- /row -->
</div> <!-- /container -->
<div>

    <div id="f">
        <div class="container">
            <div class="row">
                </br>
                </br>
            </div>
        </div>
    </div>


    <!-- Bootstrap core JavaScript
    ================================================== -->
    <!-- Placed at the end of the document so the pages load faster -->
    <script src="assets/js/bootstrap.js"></script>
  </body>
</html>
